File: frontend/src/api/searchService.js
import api from './axios';

/**
 * Search users by username (contains, case-insensitive).
 * Returns a Spring-Page of { id, username, bio, profilePicture }.
 */
export const searchUsers = (q, page = 0, size = 15) =>
    api.get('/search/users', { params: { q, page, size } });

/**
 * Search posts by title OR body (published only).
 * Returns a Spring-Page of PostResponse DTO.
 */
export const searchPosts = (q, page = 0, size = 15) =>
    api.get('/search/posts', { params: { q, page, size } });

File: src/main/java/com/catalin/vibelog/service/SearchService.java
package com.catalin.vibelog.service;

import com.catalin.vibelog.dto.response.UserResponseDTO;
import com.catalin.vibelog.dto.response.PostResponse;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

/**
 * Search operations across users and posts.
 */
public interface SearchService {

    /**
     * Search users by username fragment.
     *
     * @param q        the substring to match in usernames
     * @param pageable paging information
     * @return a page of {@link UserResponseDTO}
     */
    Page<UserResponseDTO> searchUsers(String q, Pageable pageable);

    /**
     * Search published posts by title/body fragment.
     *
     * @param q        the substring to match in title or body
     * @param pageable paging information
     * @return a page of {@link PostResponse}
     */
    Page<PostResponse> searchPosts(String q, Pageable pageable);
}
